### ISSUE TITLE
Effects Suddenly Stop Running

### ISSUE THREAD

**Issue by RyanMcDonald** [ngrx/effects#144]
I have been trying to debug this issue for a while now, but I've run out of ideas, so I'm hoping to get some help with this. For some reason, when I dispatch a `DELETE_LIGHT_GROUP` action, the Effect runs successfully, and gets to the `DELETE_LIGHT_GROUP_SUCCESS` reducer, but then no other Effects are running after that, no matter what action I dispatch.

I have subscribed to the `Actions` observable from @ngrx/effects, and it shows all of the actions being dispatched correctly, but the Effects are not catching them.

If I modify my reducer to not remove the light group object from my application state, then the problem goes away, and Effects continue running as normal. Why would my application state affect whether or not Effects stop running? Is there an error being swallowed somewhere? What am I missing?

Action:
```js
export class DeleteLightGroupAction implements Action {
  type = ActionTypes.DELETE_LIGHT_GROUP;

  constructor(public payload: LightGroup) { }
}
```

Dispatch:
```js
this.store.dispatch(new lightGroupActions.DeleteLightGroupAction(this.lightGroup))
```

Effect:
```js
@Effect() deleteLightGroup$: Observable<Action> = this.actions$
    .ofType(lightGroupActions.ActionTypes.DELETE_LIGHT_GROUP)
    .mergeMap(action => {
      const lightGroup = Object.assign({}, action.payload);

      return this.lightGroupService.deleteLightGroup(lightGroup)
        .mapTo(new lightGroupActions.DeleteLightGroupSuccessAction(lightGroup))
        .catch(error => {
          this.translateService.get('LIGHT_GROUP_EFFECTS.DELETE_LIGHT_GROUP_FAIL').subscribe(value => this.errorHelper.handleError({ message: value, error, action }));
          return Observable.of(new lightGroupActions.DeleteLightGroupFailAction(lightGroup));
        });
    });
```

Reducer:
```js
case lightGroupActions.ActionTypes.DELETE_LIGHT_GROUP_SUCCESS:
          const lightGroupId = action.payload.id;
          const entitiesWithoutDeletedLightGroup = Object.assign({}, state.entities);
          delete entitiesWithoutDeletedLightGroup[lightGroupId];

          return Object.assign({}, state, {
            ids: state.ids.filter(id => id !== lightGroupId),
            entities: entitiesWithoutDeletedLightGroup,
          });
```

**Comment by diicar** [comment]
Have you tried using switchMap instead of mergeMap in you effects ?

**Comment by RyanMcDonald** [comment]
@diicar Same results using switchMap.

**Comment by Abbraxar** [comment]
@diicar I had a very similar error.
It turned that one of my selectors had a silent error so you should check your selectors.
Maybe you have one selector which look for a deleted entity.

**Comment by devmao** [comment]
I guess your issue is related to this one #141.
If you add a catch to the deleteLightGroup stream, you will get the exception that breaks the stream itself.

**Comment by nfantone** [comment]
@devmao Thing is, even if you _do_ add a `.catch` and get the exception, what I'm seeing is that following effects won't be fired for some reason.

This might very well be by design, mind you. But from what I could gathered, it seems to be confusing people. Error handling is somewhat head-scratching, IMHO.

**Comment by lucastheisen** [comment]
@nfantone , I completely agree.  Basically, I would like to add this to the end of _every one_ of my effects:
```
        .catch((error) => {
            this.toaster.popAsync("error", "", error);
            return Observable.empty<Action>();
        });
```
Thus allowing for simplified effect processing:
```
...
        .switchMap(([action, state]: [FooAction, AppState]) => {
            return this.service.search({page: 0, query: {searchTerms: `"${action.payload}"`}})
                .map((result) => {
                    if (result.total === 1 ) {
                        return result;
                    }
                    else {
                        throw new Error(`Invalid id: ${action.payload}`);
                    }
                });
        })
        .switchMap((result) => {
...
```

**Comment by lucastheisen** [comment]
Ok, so while annoying, it appears you can avoid this by a single layer of nesting.  For example, if you:
```
    @Effect() public fooEffect= this.actions
        .ofType(ActionTypes.ADD_FOO)
        .switchMap((foo) => {
            if (state.foo[foo.id]) {
                this.toaster.popAsync("error", "", "Foo already added");
                return Observable.empty<FooResult>();
            }
            return this.service.search({page: 0, query: {searchTerms: `"${action.payload}"`}})
                .map((result) => {
                    if (result.total === 1 ) {
                        return result;
                    }
                    else {
                        throw new Error(`Invalid foo id: ${action.payload}`);
                    }
                })
                .switchMap((result) => {
                    return this.service.getFooMetadata(result.results[0]);
                })
                .withLatestFrom(this.store)
                .mergeMap(([foo, state]) => {
                    const effects: Action[] = [
                        new AddBarAction(foo, state),
                        new AddBazAction(foo, state),
                    ];
                    return Observable.from(effects);
                })
                .catch((error) => {
                    this.toaster.popAsync("error", "", error);
                    this.logger.warn(error);
                    return Observable.never<FlightSearchResult>();
                });
        });
```
Its annoying (_and ugly after the fashion of callback hell_), but it works.

**Comment by vkvarma** [comment]
Hi ,
I am able to dispatch action after doing the nesting of the catch into switchMap( as said above )
but when i receive any error into effects catch but my UI gets freeze .
Unable to do any action on UI (e.g. click on menu , click on button etc).

When i am calling http service from components rather than effects and handling error into the error block of subscribe inside component then everything is working fine.

**Comment by nghiado** [comment]
I have exactly the same issue as @RyanMcDonald  described.
I know the workaround is using catch inside switchMap but as @nfantone and @lucastheisen mentioned, it's annoying and ugly.

**Comment by vecernik** [comment]
Another breaking change with ngrx4 is that injected Actions in each effect class is not isolated as in previous version. This one-shot effect doesn't work now in ngrx4 :
  @Effect()
  private onstart$: Observable<Action> = this.actions$
    .ofType(
      'PersonalPlaybackFetched',
      'PersonalPlaylistsFetched',
      'PersonalBundlesFetched',
      'PersonalChatFetched'
    )
    .take(4)
    .skip(3)
    .map(() => new A.PersonalFetched())

This works as expected :

  @Effect()
  private onstart$: Observable<Action> = this.actions$
    .ofType(
      'PersonalPlaybackFetched',
      'PersonalPlaylistsFetched',
      'PersonalBundlesFetched',
      'PersonalChatFetched'
    )
    .distinctUntilKeyChanged('type')
    .bufferCount(4)
    .map(() => new A.PersonalFetched())

And this doesn't break the flow of actions for further effects in a class.

**Comment by JDillon522** [comment]
@lucastheisen 's example worked for me. In my case, if I had to edit a user and received an HTTP error from my API the `.catch()` block would catch it and output the error but all following attempts to edit would fail. This fixed it.

However, as others have already stated its ugly as hell and needs to be fixed proto, please.

**Comment by okonon** [comment]
@lucastheisen example worked for me

**Comment by jongunter** [comment]
Have to agree with some of the sentiment here. Error handling could use some improvement, or we could better lay down some standards for the community to use  This has been the effect pattern I've had to use in some of my apps:
```
@Effect()
    loadDetails$: Observable<Action> = this.actions$
        .ofType(details.LOAD)
        .map(toPayload)
        .switchMap(payload => {
            try {
                return this.orderService
                    .get(payload.id)
                    .map((detail: OrderDetail) => {
                        return new details.LoadSuccessAction(detail)
                    })
                    .catch(err => {
                        return Observable.of(new details.LoadFailure(err.toString()))
                    })
            } catch (err) {
                return Observable.of(new details.LoadFailure(err.toString()))
            }
        });
``` 
Notice there are **TWO** `catch`'s . The first is to catch any HTTP errors that might have occurred.  The second is to catch any errors coming from `this.orderService` before the HTTP call is made, or the error that would occur if `payload` was not defined for some reason (error when trying to access `id). I guess I could put a try/catch in  `this.orderService` and return and error-ed observable if an error occurs, but that still doesn't cover errors that could occur inside `switchMap`. 

Would a `retry` operator work here if this observable errors out? (I'm currently experimenting). The bottom line is it's getting complex and un-DRY to cover every error case.  

Maybe I'm just ranting now and I should be posting this on StackOverflow instead...

**Comment by michahell** [comment]
Effect debuggability is really bad and should be made better, it's basically trial and error now :/
